Loading the all years sparcs datset. After loading we will do the necessary preprocessing part. Everything done here is essential for modelling.

In [ ]:
from google.colab import drive
drive.mount('...')

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [ ]:
df = pd.read_parquet('/All_Sparcs_Dataset/sparcs_all_years.parquet')

In [ ]:
df.shape

(1265614, 33)

In [ ]:
df.drop_duplicates(inplace=True)

In [ ]:
df.shape

(1232243, 33)

In [ ]:
df.columns

Index(['Hospital Service Area', 'Hospital County',
       'Operating Certificate Number', 'Permanent Facility Id',
       'Facility Name', 'Age Group', 'Zip Code - 3 digits', 'Gender', 'Race',
       'Ethnicity', 'Length of Stay', 'Type of Admission',
       'Patient Disposition', 'Discharge Year', 'CCSR Diagnosis Code',
       'CCSR Diagnosis Description', 'CCSR Procedure Code',
       'CCSR Procedure Description', 'APR DRG Code', 'APR DRG Description',
       'APR MDC Code', 'APR MDC Description', 'APR Severity of Illness Code',
       'APR Severity of Illness Description', 'APR Risk of Mortality',
       'APR Medical Surgical Description', 'Payment Typology 1',
       'Payment Typology 2', 'Payment Typology 3', 'Birth Weight',
       'Emergency Department Indicator', 'Total Charges', 'Total Costs'],
      dtype='object')

In [ ]:
df.dtypes

Hospital Service Area                   object
Hospital County                         object
Operating Certificate Number           float64
Permanent Facility Id                  float64
Facility Name                           object
Age Group                               object
Zip Code - 3 digits                     object
Gender                                  object
Race                                    object
Ethnicity                               object
Length of Stay                           int64
Type of Admission                       object
Patient Disposition                     object
Discharge Year                           int64
CCSR Diagnosis Code                     object
CCSR Diagnosis Description              object
CCSR Procedure Code                     object
CCSR Procedure Description              object
APR DRG Code                             int64
APR DRG Description                     object
APR MDC Code                             int64
APR MDC Description                     object
APR Severity of Illness Code             int64
APR Severity of Illness Description     object
APR Risk of Mortality                   object
APR Medical Surgical Description        object
Payment Typology 1                      object
Payment Typology 2                      object
Payment Typology 3                      object
Birth Weight                             int64
Emergency Department Indicator          object
Total Charges                          float64
Total Costs                            float64
dtype: object

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1232243 entries, 0 to 1265613
Data columns (total 33 columns):
 #   Column                               Non-Null Count    Dtype  
---  ------                               --------------    -----  
 0   Hospital Service Area                1230805 non-null  object 
 1   Hospital County                      1230805 non-null  object 
 2   Operating Certificate Number         1230805 non-null  float64
 3   Permanent Facility Id                1230805 non-null  float64
 4   Facility Name                        1232243 non-null  object 
 5   Age Group                            1232243 non-null  object 
 6   Zip Code - 3 digits                  1213522 non-null  object 
 7   Gender                               1232243 non-null  object 
 8   Race                                 1232243 non-null  object 
 9   Ethnicity                            1232243 non-null  object 
 10  Length of Stay                       1232243 non-null  int64  
 11  Typ

In [ ]:
df.drop(columns=['Payment Typology 2','Payment Typology 3'], inplace=True)

In [ ]:
df.shape

(1232243, 31)

In [ ]:
df.drop(columns=['Patient Disposition','Discharge Year'], inplace=True)

In [ ]:
df.drop(columns=['Total Charges','Total Costs'], inplace=True)

In [ ]:
df.head()

Hospital Service Area Hospital County  Operating Certificate Number  \
0            Western NY        Allegany                      228000.0   
1            Western NY        Allegany                      228000.0   
2            Western NY        Allegany                      228000.0   
3            Western NY        Allegany                      228000.0   
4            Western NY        Allegany                      228000.0   

   Permanent Facility Id                                      Facility Name  \
0                   39.0  Memorial Hosp of Wm F & Gertrude F Jones A/K/A...   
1                   39.0  Memorial Hosp of Wm F & Gertrude F Jones A/K/A...   
2                   39.0  Memorial Hosp of Wm F & Gertrude F Jones A/K/A...   
3                   39.0  Memorial Hosp of Wm F & Gertrude F Jones A/K/A...   
4                   39.0  Memorial Hosp of Wm F & Gertrude F Jones A/K/A...   

  Age Group Zip Code - 3 digits Gender        Race          Ethnicity  ...  \
0   0 to 17                 148      F  Other Race  Not Span/Hispanic  ...   
1   0 to 17                 148      F       White            Unknown  ...   
2   0 to 17                 148      F       White  Not Span/Hispanic  ...   
3   0 to 17                 147      F       White   Spanish/Hispanic  ...   
4   0 to 17                 147      F       White            Unknown  ...   

                                 APR DRG Description APR MDC Code  \
0  Neonate birthwt >2499g, normal newborn or neon...           15   
1  Neonate birthwt >2499g, normal newborn or neon...           15   
2  Neonate birthwt >2499g, normal newborn or neon...           15   
3  Neonate birthwt >2499g, normal newborn or neon...           15   
4  Neonate birthwt >2499g, normal newborn or neon...           15   

                                 APR MDC Description  \
0  Newborns and Other Neonates with Conditions Or...   
1  Newborns and Other Neonates with Conditions Or...   
2  Newborns and Other Neonates with Conditions Or...   
3  Newborns and Other Neonates with Conditions Or...   
4  Newborns and Other Neonates with Conditions Or...   

  APR Severity of Illness Code APR Severity of Illness Description  \
0                            1                               Minor   
1                            1                               Minor   
2                            1                               Minor   
3                            1                               Minor   
4                            1                               Minor   

  APR Risk of Mortality  APR Medical Surgical Description  \
0                 Minor                           Medical   
1                 Minor                           Medical   
2                 Minor                           Medical   
3                 Minor                           Medical   
4                 Minor                           Medical   

       Payment Typology 1  Birth Weight Emergency Department Indicator  
0                Medicaid          2800                              N  
1  Blue Cross/Blue Shield          3500                              N  
2                Self-Pay          3300                              N  
3                Medicaid          3200                              N  
4                Medicaid          2600                              N  

[5 rows x 27 columns]

In [ ]:
df.shape

(1232243, 27)

In [ ]:
# verify names
print(df.columns.tolist())

['Hospital Service Area', 'Hospital County', 'Operating Certificate Number', 'Permanent Facility Id', 'Facility Name', 'Age Group', 'Zip Code - 3 digits', 'Gender', 'Race', 'Ethnicity', 'Length of Stay', 'Type of Admission', 'CCSR Diagnosis Code', 'CCSR Diagnosis Description', 'CCSR Procedure Code', 'CCSR Procedure Description', 'APR DRG Code', 'APR DRG Description', 'APR MDC Code', 'APR MDC Description', 'APR Severity of Illness Code', 'APR Severity of Illness Description', 'APR Risk of Mortality', 'APR Medical Surgical Description', 'Payment Typology 1', 'Birth Weight', 'Emergency Department Indicator']


Resolving the issue with permanent facility id and facility name

In [ ]:
# No. of Hospital County facility id
len(df['Permanent Facility Id'].unique())

142

In [ ]:
# No. of Hospital County facility name
len(df['Facility Name'].value_counts().unique())

185

Since we are not specifically studying Hospital-level effects, we can drop these columns as they add too much noise for a general clinical prediction

In [ ]:
df.drop(columns=["Permanent Facility Id","Facility Name"], inplace=True)

In [ ]:
df.shape

(1232243, 25)

Now we have to deal with the CCSR and MDC codes and descriptions

In [ ]:
df[['CCSR Diagnosis Code','CCSR Diagnosis Description']].nunique()

CCSR Diagnosis Code           237
CCSR Diagnosis Description    209
dtype: int64

there are more CCSR diagnosis descriptions than codes, which means more than 1 code sometimes mean the same description.

In [ ]:
cols = ['CCSR Diagnosis Code', 'CCSR Procedure Code','APR DRG Code','APR MDC Code','APR Severity of Illness Code']
print(df[cols].dtypes)

CCSR Diagnosis Code             object
CCSR Procedure Code             object
APR DRG Code                     int64
APR MDC Code                     int64
APR Severity of Illness Code     int64
dtype: object


The CCSR codes are often alphanumeric, so we can't convert it to int64

In [ ]:
df.groupby(['CCSR Diagnosis Code'])['CCSR Diagnosis Description'].nunique().sort_values(ascending=False)

CCSR Diagnosis Code
10        1
100       1
101       1
102       1
105       1
         ..
RSP017    1
SKN001    1
SYM001    1
SYM005    1
SYM016    1
Name: CCSR Diagnosis Description, Length: 237, dtype: int64

Dropping all the descriptions for modelling purpose

In [ ]:
df.drop(columns=['CCSR Diagnosis Description','CCSR Procedure Description','APR Severity of Illness Description',
                 'APR DRG Description','APR MDC Description'], inplace=True)

In [ ]:
df.shape

(1232243, 20)

Now we have to deal with null values of several columns of the entire dataset

In [ ]:
df.isna().sum()

Hospital Service Area                1438
Hospital County                      1438
Operating Certificate Number         1438
Age Group                               0
Zip Code - 3 digits                 18721
Gender                                  0
Race                                    0
Ethnicity                               0
Length of Stay                          0
Type of Admission                       0
CCSR Diagnosis Code                     0
CCSR Procedure Code                     0
APR DRG Code                            0
APR MDC Code                            0
APR Severity of Illness Code            0
APR Risk of Mortality                 526
APR Medical Surgical Description        0
Payment Typology 1                      0
Birth Weight                            0
Emergency Department Indicator          0
dtype: int64

First we have to deal with the 1438 null values of hospital service area and hospital county

In [ ]:
# if the 1438 rows have null hospital county, the rows are untrustworthy. Best would be to drop the rows
# See how many rows you have before cleaning
rows_before = len(df)
print(f"Number of rows before dropping: {rows_before}")

# Drop any row where 'Permanent Facility Id' is null
df.dropna(subset=['Hospital County'], inplace=True)

# See how many rows are left
rows_after = len(df)
print(f"Number of rows after dropping: {rows_after}")
print(f"Total rows removed: {rows_before - rows_after}")

Number of rows before dropping: 1232243
Number of rows after dropping: 1230805
Total rows removed: 1438


Dealing with the 3-digits

In [ ]:
df['Zip Code - 3 digits'].dtypes

dtype('O')

ZIp code is something we should not convert Integers, as these are simply designations of different places. Also there are so many nulls, so guessing the location of the patient is very difficult . We can't even one-hot encode them since these will add too many columns and make the models huge ,slow and prone to overfitting

In [ ]:
#considering the above problem, dropping them would solve the missing value problem without losing any rows

In [ ]:
df.drop(columns=['Zip Code - 3 digits'], inplace=True)

In [ ]:
df.shape

(1230805, 19)

Now dealing with APR risk of mortality

In [ ]:
print(df['APR Risk of Mortality'].dtypes)

object


In [ ]:
#Apr risk of mortality means a number like 1,2,3 etc. where the risk increases with the integer, like minor, moderate , major etc.
#if we keep some as 'NA' then it should be converted to String, the model will lose context.
#Also we cant keep it 0 since we never know the exact degree of illness
df.dropna(subset=['APR Risk of Mortality'],inplace=True)

In [ ]:
df.isna().sum()

Hospital Service Area               0
Hospital County                     0
Operating Certificate Number        0
Age Group                           0
Gender                              0
Race                                0
Ethnicity                           0
Length of Stay                      0
Type of Admission                   0
CCSR Diagnosis Code                 0
CCSR Procedure Code                 0
APR DRG Code                        0
APR MDC Code                        0
APR Severity of Illness Code        0
APR Risk of Mortality               0
APR Medical Surgical Description    0
Payment Typology 1                  0
Birth Weight                        0
Emergency Department Indicator      0
dtype: int64

In [ ]:
print(len(df))

1230280


In [ ]:
df.nunique()

Hospital Service Area                 8
Hospital County                      50
Operating Certificate Number        128
Age Group                             5
Gender                                3
Race                                  4
Ethnicity                             4
Length of Stay                      120
Type of Admission                     1
CCSR Diagnosis Code                 233
CCSR Procedure Code                 378
APR DRG Code                        198
APR MDC Code                         23
APR Severity of Illness Code          4
APR Risk of Mortality                 4
APR Medical Surgical Description      2
Payment Typology 1                   10
Birth Weight                         90
Emergency Department Indicator        2
dtype: int64

Now some more things to be done before modelling

In [ ]:
df['APR MDC Code'].value_counts()

APR MDC Code
15    1229458
14        295
4          76
5          74
23         62
1          45
19         44
8          44
6          37
18         25
10         19
13         19
11         14
9          14
20         13
3          10
2           8
16          7
7           5
24          4
21          3
17          2
25          2
Name: count, dtype: int64

In [ ]:
df['Type of Admission'].value_counts()

Type of Admission
Newborn    1230280
Name: count, dtype: int64

In [ ]:
df['Age Group'].value_counts()

Age Group
0 to 17        1229667
30 to 49           237
18 to 29           155
50 to 69           130
70 or Older         91
Name: count, dtype: int64

1.   Though the type of admission of all of them is newborn, still we can see the MDC 15 code has majority of the data.
2.   In rare cases, a baby is born in the hospital (Admission = Newborn), but their primary medical problem is so specific or severe that it gets grouped into a different Major Diagnostic Category (MDC).
Example A (MDC 15): A baby is born premature or with jaundice. This is a standard neonatal problem. They go to MDC 15 ("Newborns and Other Neonates"). This is your target population.
Example B (MDC 01 - Nervous System): A baby requires immediate, complex brain surgery. They might be grouped into MDC 01 (Nervous System).


1.   This MDC code might result in very long stay and also very high cost. So they are massive potential outliers for our task. So it's better to drop the other MDC code and retain the 15 only
2. Also the age group have some of the mother's age , as we found in the EDA file, so it's better to drop the entiire column, and MDC 15 will take care of rest




In [ ]:
#drop age group
df.drop(columns=['Age Group'], inplace=True)

In [ ]:
#filter MDC 15
df = df[df['APR MDC Code']==15]
print(f"Population refined to clinical newborns, no. of rows are: {len(df)}")

Population refined to clinical newborns, no. of rows are: 1229458


In [ ]:
print(df.columns.tolist())

['Hospital Service Area', 'Hospital County', 'Operating Certificate Number', 'Gender', 'Race', 'Ethnicity', 'Length of Stay', 'Type of Admission', 'CCSR Diagnosis Code', 'CCSR Procedure Code', 'APR DRG Code', 'APR MDC Code', 'APR Severity of Illness Code', 'APR Risk of Mortality', 'APR Medical Surgical Description', 'Payment Typology 1', 'Birth Weight', 'Emergency Department Indicator']


In [ ]:
df.nunique()

Hospital Service Area                 8
Hospital County                      50
Operating Certificate Number        124
Gender                                3
Race                                  4
Ethnicity                             4
Length of Stay                      120
Type of Admission                     1
CCSR Diagnosis Code                  68
CCSR Procedure Code                 347
APR DRG Code                         30
APR MDC Code                          1
APR Severity of Illness Code          4
APR Risk of Mortality                 4
APR Medical Surgical Description      2
Payment Typology 1                   10
Birth Weight                         90
Emergency Department Indicator        2
dtype: int64

In [ ]:
#we should drop the MDC and type of admission col, since they have a single value for each row,i.e zero variance. The model will not learn anything from these
df.drop(columns=['APR MDC Code', 'Type of Admission'], inplace=True)

In [ ]:
df.shape

(1229458, 16)

In [ ]:
#also the operating certificate no. has 124 unique values, of we encode them, it will add these many cols, this might overfit the model
#we want a model that predicts based on the baby's health, not based on which specific building they are in, so dropping the location identifiers

df.drop(columns=['Operating Certificate Number','Hospital County','Hospital Service Area'], inplace=True)

In [ ]:
df.nunique()

Gender                                3
Race                                  4
Ethnicity                             4
Length of Stay                      120
CCSR Diagnosis Code                  68
CCSR Procedure Code                 347
APR DRG Code                         30
APR Severity of Illness Code          4
APR Risk of Mortality                 4
APR Medical Surgical Description      2
Payment Typology 1                   10
Birth Weight                         90
Emergency Department Indicator        2
dtype: int64

In [ ]:
df['Length of Stay'].value_counts()

Length of Stay
2      646466
3      255548
1      160400
4       61296
5       18810
        ...  
118        48
117        44
119        41
115        40
116        37
Name: count, Length: 120, dtype: int64

A Length of Stay (LOS) of 116 days is clinically valid for MDC 15.

2-4 Days: Healthy, full-term newborn.

10-30 Days: Moderate prematurity or infection.

60-120 Days: Extreme prematurity (micro-preemies) or severe congenital defects requiring long-term NICU care.

The standard data science solution is to keep the "Long Tail" (up to 99%) but cut off the "Extreme Anomalies" (the top 1%).

In [ ]:
#calculating the 99 percentile
cutoff_val = df['Length of Stay'].quantile(0.99)

print(f"99% of babies stay is {cutoff_val} days or less")
print(f"Max stay in the dataset currently is {df['Length of Stay'].max()}")

total_outliers = len(df[df['Length of Stay']>cutoff_val])
print(f"dropping {total_outliers} rows with los > {cutoff_val}")

99% of babies stay is 42.0 days or less
Max stay in the dataset currently is 120
dropping 12077 rows with los > 42.0


In [ ]:
df = df[df['Length of Stay']<=42]

In [ ]:
df['Length of Stay'].value_counts()

Length of Stay
2     646466
3     255548
1     160400
4      61296
5      18810
6      11314
7       9710
8       6361
9       4966
10      4312
11      3772
12      3165
13      2943
14      2513
15      2296
16      2035
17      1852
18      1635
19      1474
20      1401
21      1304
22      1226
23      1037
24      1012
25       893
27       829
26       820
28       699
31       667
29       666
30       654
32       600
33       570
35       569
34       555
36       537
37       468
38       443
39       404
40       402
41       384
Name: count, dtype: int64

In [ ]:
df.columns

Index(['Gender', 'Race', 'Ethnicity', 'Length of Stay', 'CCSR Diagnosis Code',
       'CCSR Procedure Code', 'APR DRG Code', 'APR Severity of Illness Code',
       'APR Risk of Mortality', 'APR Medical Surgical Description',
       'Payment Typology 1', 'Birth Weight', 'Emergency Department Indicator'],
      dtype='object')

In [ ]:
#export upto this
df.to_parquet('/All_Sparcs_Dataset/sparcs_all_years_3.parquet')

Now we have to one-hot encode the categorical values

In [ ]:
df = pd.read_parquet('/All_Sparcs_Dataset/sparcs_all_years_3.parquet')

In [ ]:
#cols to one-hot encode
categorical_cols = ['Gender','Race','Ethnicity','Payment Typology 1','Emergency Department Indicator','APR Medical Surgical Description','CCSR Diagnosis Code','CCSR Procedure Code','APR DRG Code']

print("Encoding the data, this will add may new columns")
df = pd.get_dummies(df, columns = categorical_cols, drop_first=True)

Encoding the data, this will add may new columns


In [ ]:
df.shape

(1217008, 441)

In [ ]:
print(df.dtypes)

Length of Stay                   int64
APR Severity of Illness Code     int64
APR Risk of Mortality           object
Birth Weight                     int64
Gender_M                          bool
                                 ...  
APR DRG Code_633                  bool
APR DRG Code_634                  bool
APR DRG Code_636                  bool
APR DRG Code_639                  bool
APR DRG Code_640                  bool
Length: 441, dtype: object


In [ ]:
df['APR Risk of Mortality'].value_counts()

APR Risk of Mortality
Minor       1189045
Moderate      19651
Major          5945
Extreme        2367
Name: count, dtype: int64

In [ ]:
#converting the risk of mortality col to int64
mapping_dict = {
    'Minor': 1,
    'Moderate': 2,
    'Major': 3,
    'Extreme': 4
}
df['APR Risk of Mortality'] = df['APR Risk of Mortality'].replace(mapping_dict)

In [ ]:
print(df.dtypes)

Length of Stay                  int64
APR Severity of Illness Code    int64
APR Risk of Mortality           int64
Birth Weight                    int64
Gender_M                         bool
                                ...  
APR DRG Code_633                 bool
APR DRG Code_634                 bool
APR DRG Code_636                 bool
APR DRG Code_639                 bool
APR DRG Code_640                 bool
Length: 441, dtype: object


Scaling the birthweight column

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [ ]:
# we have to scale the birthweight of the train set and test set seperately,
#if the whole dataset is scaled at once the train set will peep into the test set(data leakage)
target = 'Length of Stay'
col_to_scale = ['Birth Weight']

In [ ]:
#train-test split
X = df.drop(columns=[target])
y = df[target]
# splitting data int train 80% and test 20%
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
#scaling
scaler = StandardScaler()
# Use .loc to avoid SettingWithCopy warnings
X_train.loc[:, col_to_scale] = scaler.fit_transform(X_train[col_to_scale])
X_test.loc[:, col_to_scale] = scaler.transform(X_test[col_to_scale])

In [ ]:
print("DATA READY!")
print(f"X_train shape: {X_train.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"X_test shape:  {X_test.shape}")
print(f"y_test shape:  {y_test.shape}")
print("-" * 30)

✅ DATA READY!
X_train shape: (973606, 440)
y_train shape: (973606,)
X_test shape:  (243402, 440)
y_test shape:  (243402,)
------------------------------


Now the modelling part

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import time

In [ ]:
# --- HELPER FUNCTION TO PRINT RESULTS ---
def print_results(model_name, y_true, y_pred, training_time):
    #rmse = mean_squared_error(y_true, y_pred, squared=False)
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)  # Calculate RMSE manually
    mae = mean_absolute_error(y_true, y_pred)
    r2 = r2_score(y_true, y_pred)

    print(f"\n {model_name} RESULTS:")
    print(f"Time to Train: {training_time:.2f} seconds")
    print(f"RMSE: {rmse:.4f}  <-- THE SCORE TO BEAT")
    print(f"MAE:  {mae:.4f}")
    print(f"R²:   {r2:.4f}")
    return rmse

In [ ]:
# 1. LINEAR REGRESSION
# ==========================================
print("Training Linear Regression...")
start_time = time.time()

lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

lr_pred = lr_model.predict(X_test)
lr_time = time.time() - start_time
lr_rmse = print_results("Linear Regression", y_test, lr_pred, lr_time)

Training Linear Regression...

📊 Linear Regression RESULTS:
Time to Train: 58.53 seconds
RMSE: 2.2159  <-- THE SCORE TO BEAT
MAE:  1.0321
R²:   0.6182


In [ ]:
# 2. RANDOM FOREST
# ==========================================
print("-" * 30)
print("Training Random Forest... (This uses all CPU cores, please wait)")
start_time = time.time()

# n_jobs=-1 forces the code to use every processor core you have for speed
rf_model = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
rf_model.fit(X_train, y_train)

rf_pred = rf_model.predict(X_test)
rf_time = time.time() - start_time
rf_rmse = print_results("Random Forest", y_test, rf_pred, rf_time)

------------------------------
Training Random Forest... (This uses all CPU cores, please wait)

📊 Random Forest RESULTS:
Time to Train: 1972.34 seconds
RMSE: 2.1787  <-- THE SCORE TO BEAT
MAE:  0.9769
R²:   0.6309


In [ ]:
print("-" * 30)
if rf_rmse < lr_rmse:
    print(f"✅ WINNER: Random Forest is your official Baseline.")
    print(f"TARGET RMSE TO BEAT: {rf_rmse:.4f}")
else:
    print(f"✅ WINNER: Linear Regression is your official Baseline.")
    print(f"TARGET RMSE TO BEAT: {lr_rmse:.4f}")

------------------------------
✅ WINNER: Random Forest is your official Baseline.
TARGET RMSE TO BEAT: 2.1787
